tensorflow_addons/optimizers/novograd.py

```diff
@@ -0,0 +1,246 @@
+                one_minus_beta_1_t=1 - beta_1_t,
+            ))
+
+    def set_weights(self, weights):
+        params = self.weights
+        # If the weights are generated by Keras V1 optimizer, it includes vhats
+        # even without amsgrad, i.e, V1 optimizer has 3x + 1 variables, while V2
+        # optimizer has 2x + 1 variables. Filter vhats out for compatibility.
+        num_vars = int((len(params) - 1) / 2)
+        if len(weights) == 3 * num_vars + 1:
+            weights = weights[:len(params)]
+        super(Novograd, self).set_weights(weights)
+
+    def _resource_apply_dense(self, grad, var, apply_state=None):
+        var_device, var_dtype = var.device, var.dtype.base_dtype
+        coefficients = ((apply_state or {}).get((var_device, var_dtype))
+                        or self._fallback_apply_state(var_device, var_dtype))
+        weight_decay = self._get_hyper('weight_decay')
+        grad_averaging = self._get_hyper('grad_averaging')
+
+        v = self.get_slot(var, 'v')
+        g_2 = tf.reduce_sum(tf.square(tf.cast(grad, tf.float32)))
+        v_t = tf.cond(
+            tf.equal(self.iterations,
+                     0), lambda: g_2, lambda: v * coefficients['beta_2_t'] +
+            g_2 * coefficients['one_minus_beta_2_t'])
+        v_t = v.assign(v_t, use_locking=self._use_locking)
+
+        if self.amsgrad:
+            vhat = self.get_slot(var, 'vhat')
+            vhat_t = vhat.assign(
+                tf.maximum(vhat, v_t), use_locking=self._use_locking)
+            grad = grad / (tf.sqrt(vhat_t) + self.epsilon)
+        else:
+            grad = grad / (tf.sqrt(v_t) + self.epsilon)
+        grad = tf.cond(
+            tf.greater(weight_decay,
+                       0), lambda: grad + weight_decay * var, lambda: grad)
+        grad = tf.cond(grad_averaging, lambda: grad * coefficients[
+            'one_minus_beta_1_t'], lambda: grad)
+        m = self.get_slot(var, 'm')
+        return training_ops.resource_apply_keras_momentum(
+            var.handle,
+            m.handle,
+            coefficients['lr_t'],
+            grad,
+            coefficients['beta_1_t'],
+            use_locking=self._use_locking,
+            use_nesterov=False)
+
+    def _resource_apply_sparse(self, grad, var, indices, apply_state=None):
+        var_device, var_dtype = var.device, var.dtype.base_dtype
+        coefficients = ((apply_state or {}).get((var_device, var_dtype))
+                        or self._fallback_apply_state(var_device, var_dtype))
+        weight_decay = self._get_hyper('weight_decay')
+        grad_averaging = self._get_hyper('grad_averaging')
+
+        v = self.get_slot(var, 'v')
+        g_2 = tf.reduce_sum(tf.square(tf.cast(grad, tf.float32)))
+        # v is just a scalar and does not need to involve sparse tensors.
+        v_t = tf.cond(
+            tf.equal(self.iterations,
+                     0), lambda: g_2, lambda: v * coefficients['beta_2_t'] +
```

Comment:
Is it possible to put `tf.equal(iterations, 0)` in one line, which is more readable.

Reply:
Done.